propagation_ot.go

```diff
@@ -191,66 +169,39 @@ func (p *binaryPropagator) Join(
 	if !ok {
 		return nil, opentracing.ErrInvalidCarrier
 	}
-	// Handle the trace, span ids, and sampled status.
-	var traceID, propagatedSpanID int64
-	var sampledByte byte
 
-	if err := binary.Read(carrier, binary.BigEndian, &traceID); err != nil {
-		if err == io.EOF {
-			return nil, opentracing.ErrTraceNotFound
-		}
-		return nil, opentracing.ErrTraceCorrupted
-	}
-	if err := binary.Read(carrier, binary.BigEndian, &propagatedSpanID); err != nil {
+	// Read the length of marshalled binary. io.ReadAll isn't that performant
+	// since it keeps resizing the underlying buffer as it encounters more bytes
+	// to read. By reading the length, we can allocate a fixed sized buf and read
+	// the exact amount of bytes into it.
+	var len uint32
+	if err := binary.Read(carrier, binary.BigEndian, &len); err != nil {
```

Comment:
Only caveat here is that we want these encodings to work across languages and many platforms do not have nice `binary.ReadUVarint` encoding/decoding intrinsically.

Meta-caveat would be that we are no binding ourselves to platforms that at least have protobuf support, at which point they must *somehow* have access to consistent varint encoding here.

... But this brings up a good point. Since we need this to be portable across platforms and we are already binding to protobufs, better to use something like https://github.com/gogo/protobuf/blob/200875106f3bf0eb01eb297dae30b250a25ffc84/proto/decode.go#L63 directly?